hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/multipart/MultipartDownloader.java

```diff
@@ -0,0 +1,171 @@
+                            currentOffset += chunk.length;
+                        }
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        throw new RuntimeException(e);
+                    } finally {
+                        if (abortableInputStream != null) {
+                            try {
+                                if (isAbort.get()) {
+                                    abortableInputStream.abort();
+                                } else {
+                                    abortableInputStream.close();
+                                }
+                            } catch (IOException e) {
+                                // TODO use try-with-resources
+                                throw new RuntimeException(e);
+                            }
+                        }
+                    }
+                }
+            });
+
+            partFutures.add(partFuture);
+        }
+
+        for (ListenableFuture<?> partFuture : partFutures) {
+            Futures.addCallback(partFuture, new FutureCallback<Object>() {
+                @Override
+                public void onSuccess(Object o) {
+
+                }
+
+                @Override
+                public void onFailure(Throwable throwable) {
+                    orderingQueue.closeWithException(new RuntimeException("Exception caught while downloading part", throwable));
+                    cancelAllDownloads(partFutures);
+                }
+            }, DirectExecutor.INSTANCE);
+        }
+
+        return new OrderingQueueInputStream(orderingQueue, new Runnable() {
+            @Override
+            public void run() {
+                isAbort.set(false);
+                cancelAllDownloads(partFutures);
+            }
+        }, new Runnable() {
+            @Override
+            public void run() {
+                isAbort.set(true);
+                cancelAllDownloads(partFutures);
+            }
+        });
+    }
+
+    private void cancelAllDownloads(List<ListenableFuture<?>> partFutures) {
+        for (ListenableFuture<?> partFuture : partFutures) {
+            partFuture.cancel(true);
+        }
+    }
+
+    /**
+     * Copied from Guava's MoreExecutors#directExecutor.
+     * <p>
+     * Had to copy this because directExecutor doesn't exist in Guava 11.0, and sameThreadExecutor
+     * from Guava 11.0 doesn't exist in newer versions of Guava.
+     */
+    private enum DirectExecutor implements Executor {
+        INSTANCE;
+
+        @Override
+        public void execute(Runnable command) {
+            command.run();
+        }
+
+        @Override
+        public String toString() {
+            return "MultipartDownloader.directExecutor()";
+        }
+    }
+
+    /**
+     * Copied from Math#toIntExact from Java 7.
```

Comment:
would suggest using Guava's `Ints.checkCast` instead of copying

Reply:
Done